### PR TITLE
domain,executor: store topN slow query in domain

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Store topN slow query in domain, so later we can retrieve it later.

### What is changed and how it works?

logSlowQuery in session will send a copy to domain, the domain maintains a heap to store the 
topN recent slow queries.

### Check List 

Tests 

- Unit test

### CODE REVIEW

**Comment by shenli** [comment]
Why put it in the `domain`?

**Comment by winkyao** [comment]
If put it in domain, it may lead to cycle import?

**Comment by tiancaiamao** [comment]
There is no cycle import. `executor` imports domain, `domain` doesn't imports `executor`. @winkyao 

So if we don't put it in `domain`, where should we put it? @shenli

File: domain/domain.go [hunk 1]
```diff
@@ -471,6 +514,7 @@ func NewDomain(store kv.Storage, ddlLease time.Duration, statsLease time.Duratio
 		sysSessionPool:  pools.NewResourcePool(factory, capacity, capacity, resourceIdleTimeout),
 		statsLease:      statsLease,
 		infoHandle:      infoschema.NewHandle(store),
+		slowQuery:       newTopNSlowQuery(30, time.Hour*24*7),
```
Comment (coocood) [inline domain/domain.go]: This should be configurable.

File: domain/topn_slow_query.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+			if q.data[i].duration <= q.data[smaller].duration {
+				break
+			}
+			q.data[i], q.data[smaller] = q.data[smaller], q.data[i]
+			i = smaller
+		}
+	}
+}
+
+func (q *topNSlowQuery) shiftUp(end int) {
```
Comment (coocood) [inline domain/topn_slow_query.go]: siftUp?

**Comment by coocood** [comment]
This is not a performance hot spot, we can just use heap in the standard library.

**Comment by tiancaiamao** [comment]
PTAL @coocood  @winkyao

File: domain/topn_slow_query.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+// topNSlowQuery maintains a heap to store recent slow queries.
+// N = 30, recent = 7 days by default.
+type topNSlowQuery struct {
```
Comment (coocood) [inline domain/topn_slow_query.go]: There are multiple query entries, so I think `topNSlowQueries` is better.

File: domain/topn_slow_query.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+		recent: recent,
+		ch:     make(chan *slowQueryInfo, 1000),
+	}
+}
+
+func (q *topNSlowQuery) Close() {
+	close(q.ch)
+}
+
+func (q *topNSlowQuery) Push(info *slowQueryInfo) {
```
Comment (coocood) [inline domain/topn_slow_query.go]: Why not implement the `Heap` interface in this type?
For the name confliction, we can change this `Push` to `Add` or `Append`.

**Comment by tiancaiamao** [comment]
PTAL @coocood

File: executor/adapter.go [hunk 5]
```diff
@@ -371,6 +372,13 @@ func (a *ExecStmt) logSlowQuery(txnTS uint64, succ bool) {
 		logutil.SlowQueryLogger.Warnf(
 			"[SLOW_QUERY] %vcost_time:%v %s succ:%v con:%v user:%s txn_start_ts:%v database:%v %v%vsql:%v",
 			internal, costTime, sessVars.StmtCtx.GetExecDetails(), succ, connID, user, txnTS, currentDB, tableIDs, indexIDs, sql)
+		if !sessVars.InRestrictedSQL {
```
Comment (winkyao) [inline executor/adapter.go]: Just log general sql? I prefer to keep two heap to log the general sql and internal sql.

File: domain/domain.go [hunk 6]
```diff
@@ -329,6 +331,46 @@ func (do *Domain) Reload() error {
...
+		db:       db,
+		tableIDs: tableIDs,
+		indexIDs: indexIDs,
+	}:
+	default:
+	}
+}
+
+func (do *Domain) topNSlowQueryLoop() {
+	defer do.wg.Done()
```
Comment (winkyao) [inline domain/domain.go]: recover this goroutine.

File: domain/topn_slow_query.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+			q.data[idx] = q.data[i]
+			idx++
+		}
+	}
+	if len(q.data) == idx {
+		return
+	}
+
+	// Rebuild the heap.
+	q.data = q.data[:idx]
```
Comment (winkyao) [inline domain/topn_slow_query.go]: You must use a lock to protect q.data, as long as you need to read the slice later.

File: domain/topn_slow_query.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+func (q *topNSlowQueries) Len() int           { return len(q.data) }
+func (q *topNSlowQueries) Less(i, j int) bool { return q.data[i].duration < q.data[j].duration }
+func (q *topNSlowQueries) Swap(i, j int)      { q.data[i], q.data[j] = q.data[j], q.data[i] }
+
+func (q *topNSlowQueries) Push(x interface{}) {
+	q.data = append(q.data, x.(*slowQueryInfo))
+}
+
+func (q *topNSlowQueries) Pop() interface{} {
```
Comment (winkyao) [inline domain/topn_slow_query.go]: Pop can only return the minimum duration query, how can we implement topn and how can we just peek the heap?

Comment (winkyao) [reply domain/topn_slow_query.go]: Maybe a b-tree is better?

Comment (coocood) [reply domain/topn_slow_query.go]: The read operation is not implemented in this PR.
Reading doesn't need to call Pop()

File: domain/topn_slow_query.go (hunk 7 above)
Comment (coocood) [reply domain/topn_slow_query.go]: We can do reading in the same goroutine.

Comment (tiancaiamao) [reply domain/topn_slow_query.go]: I'll use copy on read, and they will be in one goroutine, no lock. @winkyao

File: domain/topn_slow_query.go (hunk 8 above)
Comment (tiancaiamao) [reply domain/topn_slow_query.go]: Read will not be a frequent operation, so we just copy on read.
Get top 3 query is easy, copy the origin heap, Pop Pop Pop.

**Comment by tiancaiamao** [comment]
PTAL @shenli @winkyao

**Comment by coocood** [comment]
LGTM

File: domain/domain.go [hunk 9]
```diff
@@ -329,6 +331,48 @@ func (do *Domain) Reload() error {
 	return nil
 }
 
+// LogTopNSlowQuery keeps topN recent slow queries in domain.
+func (do *Domain) LogTopNSlowQuery(sql string, start time.Time, duration time.Duration,
```
Comment (zz-jason) [inline domain/domain.go]: This function takes so many parameters, which makes it hard to read and maintain, could you extract a struct to store all the parameters and pass the struct as the parameter instead?

Comment (lysu) [reply domain/domain.go]: maybe `slowQueryInfo` is :dog:

Comment (zz-jason) [reply domain/domain.go]: how about exporting `slowQueryInfo` and use `slowQueryInfo` instead?

File: domain/topn_slow_query.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+
+	"github.com/pingcap/tidb/util/execdetails"
+)
+
+type slowQueryHeap struct {
+	data []*slowQueryInfo
+}
+
+func (h *slowQueryHeap) Len() int           { return len(h.data) }
+func (h *slowQueryHeap) Less(i, j int) bool { return h.data[i].duration < h.data[j].duration }
```
Comment (zz-jason) [inline domain/topn_slow_query.go]: should it be `h.data[i].duration > h.data[j].duration`

Comment (tiancaiamao) [reply domain/topn_slow_query.go]: It's `Less` ... @zz-jason

Comment (zz-jason) [reply domain/topn_slow_query.go]: use heap[len(heap)-1] stores the slowest query?

Comment (coocood) [reply domain/topn_slow_query.go]: `heap[0]` is the fastest slow query, `heap[len(heap)-1]` may not be the slowest query.

**Comment by tiancaiamao** [comment]
PTAL @zz-jason

**Comment by zz-jason** [comment]
LGTM

**Pull request merged**